jina/executors/indexers/cache.py

```diff
@@ -22,35 +26,104 @@ def post_init(self):
 
 
 class DocIDCache(BaseCache):
-    """Store doc ids in a int64 set and persistent it to a numpy array """
+    """A key-value indexer that specializes in caching.
+    Serializes the cache to file
+    """
+
+    class CacheHandler:
+        def __init__(self, path, logger):
+            self.path = path
+            try:
+                # TODO maybe mmap?
+                self.ids = pickle.load(open(path + '.ids', 'rb'))
+                self.cache = pickle.load(open(path + '.cache', 'rb'))
+            except FileNotFoundError as e:
+                logger.warning(f'File path did not exist : {path}.ids or {path}.cache: {repr(e)}. Creating new CacheHandler...')
+                self.ids = []
```

Comment:
why one is named ids and the other cache? If this is itself a 'cache'?

Reply:
It's sort of a KV store, with one being the keys (`.ids`), and the other being the values (`.cache`). The first one contains the `ids` of the documents, the other contains the `.content_hash`-es. It is what I discussed here https://github.com/jina-ai/jina/pull/1510#discussion_r546762157

Reply:
Two possible improvements:

- using `np.mmap`. What do you think?
- if the field we cache by is `id` we don't need to use the `.cache` as a separate store. This one I'll do for sure now.

Reply:
I would then call one ids and the other hashes better?

Reply:
I would not do early optimization. I think the complete CRUD is complex enough, I would have the complete thing properly tested, and later think of potential improvements, what do you think?

Reply:
Sure, then apart from the workspace issue, what else needs to be done?

The tests in `test_cache_driver` are unit tests.

One integration test is the incremental indexing. They are all passing.

Reply:
Good, I think is really good job, I am sure we may find some nuances when integrating all together, but besides the comment on the `executor-driver-logic` I think is good